[RFC] cleanup register handling and fix missing clipboard support in `:redir`

bfredl opened the pull request:
Remove static `y_current` and `y_append` and pass `yankreg`s as arguments/return values instead. Eliminates some code of this style:

```
  struct yankreg *curr = y_current;
  // Set it to 'y_current' since 'free_yank_all' operates on it
  y_current = reg;
  free_yank_all();
  // Restore 'y_current'
  y_current = curr;
```

Also make paste in visual mode (`viwp` and such) simpler and avoid spurious clipboard provider calls when `clipboard=unnamed`. 

Some more work to be done like cleaning up bools and function docs.

bfredl committed:
ops.c: eliminate unneccesary static variable `y_append`

bfredl committed:
clipboard: simplify handling of of put in visual mode.

When clipboard=unnamed and put over visual selection, reduces number of
provider calls from 6 to 2. Also add test.

Pyrohh commented on src/nvim/ops.c:
```diff
@@ -1033,16 +1005,16 @@ do_execreg (
      */
     put_reedit_in_typebuf(silent);
-    for (i = y_current->y_size; --i >= 0; ) {
+    for (i = reg->y_size; --i >= 0; ) {
```
maybe `for (i = (reg->y_size - 1); i >= 0; i--)`

bfredl replied:
Yes, explicit `reg->y_size - 1` is clearer

Pyrohh commented on src/nvim/ops.c:
```diff
@@ -1156,17 +1128,17 @@ insert_reg (
     else {
-      for (i = 0; i < y_current->y_size; ++i) {
-        stuffescaped(y_current->y_array[i], literally);
+      for (i = 0; i < reg->y_size; ++i) {
```
`++i` should be postfix according to style guide.

bfredl replied:
:+1:

Pyrohh commented on src/nvim/ops.c:
```diff
@@ -790,61 +780,45 @@ void get_yank_register(int regname, int mode)
+    copy->y_array = NULL;
+  } else {
+    copy->y_array = xmalloc(copy->y_size * sizeof(char_u *));
+    for (linenr_T i = 0; i < copy->y_size; ++i) {
```
`++i` -> `i++`

Pyrohh commented on src/nvim/ops.c:
```diff
@@ -1303,20 +1275,20 @@ cmdline_paste_reg (
 
-  for (i = 0; i < y_current->y_size; ++i) {
-    cmdline_paste_str(y_current->y_array[i], literally);
+  for (i = 0; i < reg->y_size; ++i) {
```
`++i` -> `i++`

Pyrohh commented on src/nvim/ops.c:
```diff
@@ -4743,14 +4714,14 @@ void *get_reg_contents(int regname, int flags)
     list_T *list = list_alloc();
-    for (int i = 0; i < y_current->y_size; ++i) {
-      list_append_string(list, y_current->y_array[i], -1);
+    for (int i = 0; i < reg->y_size; ++i) {
```
`++i` -> `i++`

Pyrohh commented on src/nvim/ops.c:
```diff
@@ -4760,13 +4731,13 @@ void *get_reg_contents(int regname, int flags)
   size_t len = 0;
-  for (i = 0; i < y_current->y_size; ++i) {
-    len += STRLEN(y_current->y_array[i]);
+  for (i = 0; i < reg->y_size; ++i) {
```
`++i` -> `i++`

Pyrohh commented on src/nvim/ops.c:
```diff
@@ -4776,15 +4747,15 @@ void *get_reg_contents(int regname, int flags)
   len = 0;
-  for (i = 0; i < y_current->y_size; ++i) {
-    STRCPY(retval + len, y_current->y_array[i]);
+  for (i = 0; i < reg->y_size; ++i) {
```
`++i` -> `i++`

bfredl committed:
clipboard: cleanup `valid_yank_reg` and use it for `:redir`

This fixes missing clipboard support for `:redir`

Helped-By: Michael Reed <m.reed@mykolab.com>
Helped-By: Scott Prager <splinterofchaos@gmail.com>

bfredl committed:
ops.c: eliminate static variable `y_current`

This variable isn't stateful, and should be passed around instead.

Helped-By: Scott Prager <splinterofchaos@gmail.com>
Helped-By: Michael Reed <m.reed@mykolab.com>

bfredl commented:
Thanks, updated (still needs test for #1396, so not ready yet)

bfredl commented:
Updated with test for #1396

splinterofchaos commented on src/nvim/ops.c:
```diff
@@ -723,15 +714,13 @@ char_u *get_expr_line_src(void)
+///
+/// @param regname name of register
+/// @param writing allow only writable registers
+bool valid_yank_reg(int regname, bool writing)
```
Should say `Checks if`; missing `@return(s)`.

bfredl replied:
:+1:

splinterofchaos commented on src/nvim/ops.c:
```diff
@@ -790,61 +778,45 @@ void get_yank_register(int regname, int mode)
- */
-void put_register(int name, void *reg)
+/// return copy of contents in register `name`
+/// for use in do_put. Should be freed by caller.
```
Returns *a* copy

bfredl replied:
:+1:

splinterofchaos commented on src/nvim/ops.c:
```diff
@@ -2361,11 +2324,29 @@ static void free_yank_all(void)
  * Return FAIL for failure, OK otherwise.
  */
-int op_yank(oparg_T *oap, int deleting, int mess)
+int op_yank(oparg_T *oap, bool mess) {
```
Open brace should be on the next line. `FUNC_ATTR_NONNUL_ALL`. Also, looks like it can return a bool.

bfredl replied:
Is it a generic refactor that `int`/`OK`/`FAIL` should become stdbool ? Then perhaps it should be added to the style guide. (Right now it says `int` should be used for "error codes")

splinterofchaos replied:
Yeah... and maybe it should specify "non-boolean error codes".

bfredl replied:
The "enum" seems to actually be `OK`/`FAIL`/`NOTDONE` but `NOTDONE` is only used in one return value in `edit.c` and twice in `eval.c` (just internally in the parsing), I suppose the rest could be refactored to use true/false ( i e every case where `NOTDONE` not possible)

splinterofchaos commented on src/nvim/ops.c:
```diff
@@ -5329,9 +5284,9 @@ int get_default_register_name(void)
+/// @returns the yankreg that should be written into, or `NULL`
 /// if the register isn't a clipboard or provider isn't available.
-static struct yankreg* adjust_clipboard_name(int *name, bool quiet, bool writing) {
+static yankreg_T* adjust_clipboard_name(int *name, bool quiet) {
```
Brace should be on the next line.

bfredl replied:
:+1:

bfredl commented:
Thanks, updated to comments. I also removed some either redudant or dead error handling code. (A yank in `vim` could fail on OOM, but as we abort on OOM,)

splinterofchaos commented:
LGTM :+1:

bfredl commented:
Squashed and updated the documentation of `op_yank`. I will merge if tests pass (and no more comments)

Pyrohh commented on src/nvim/ops.c:
```diff
@@ -790,61 +778,45 @@ void get_yank_register(int regname, int mode)
+  if (copy->y_size == 0) {
+    copy->y_array = NULL;
+  } else {
+    copy->y_array = xmalloc(copy->y_size * sizeof(char_u *));
```
Could the multiplication overflow if the register contains a huge amount of text? I'm asking because I haven't checked elsewhere.

bfredl replied:
If someone yanks more than 1 billion lines on a 32 bit cpu, yes, but certainly there would be OOM before having the chance to yank anything :P But I will change to `xcalloc` anyway as it's clearer.

Pyrohh commented on src/nvim/ops.c:
```diff
@@ -5356,18 +5289,15 @@ static struct yankreg* adjust_clipboard_name(int *name, bool quiet, bool writing
 }
 
-static bool get_clipboard(int name, struct yankreg** target, bool quiet)
+static bool get_clipboard(int name, yankreg_T** target, bool quiet)
```
pointer alignment

bfredl replied:
:+1: I will `/\>\* \<` to be sure.

Pyrohh commented on src/nvim/ops.c:
```diff
@@ -5348,7 +5281,7 @@ static struct yankreg* adjust_clipboard_name(int *name, bool quiet, bool writing
       return NULL;
     }
-    struct yankreg* target;
+    yankreg_T* target;
```
pointer alignment

Pyrohh commented on src/nvim/ops.c:
```diff
@@ -5329,9 +5261,10 @@ int get_default_register_name(void)
+/// @returns the yankreg that should be written into, or `NULL`
 /// if the register isn't a clipboard or provider isn't available.
-static struct yankreg* adjust_clipboard_name(int *name, bool quiet, bool writing) {
+static yankreg_T* adjust_clipboard_name(int *name, bool quiet)
```
pointer alignment

Pyrohh commented on src/nvim/ops.c:
```diff
@@ -5356,18 +5289,15 @@ static struct yankreg* adjust_clipboard_name(int *name, bool quiet, bool writing
+static bool get_clipboard(int name, yankreg_T** target, bool quiet)
 {
-  struct yankreg* reg = adjust_clipboard_name(&name, quiet, false);
+  yankreg_T* reg = adjust_clipboard_name(&name, quiet);
```
pointer alignment

bfredl commented:
Thanks @Pyrohh and @splinterofchaos for the reviews! Updated.

bfredl commented:
qb says "1 error" but clicking on "1 error" shows nothing!

bfredl commented:
Merging, as failure is unrelated.

elmart commented:
IMHO, we shouldn't merge until being able to check until build passes. Even if failure is unrelated (in principle), we could be inadvertently composing bugs. Better to wait, we have no hurry.

bfredl commented:
Well it was written on some other PR that this failure could be disregarded for now. (After all it is a point on the build matrix that was missing, and not a "row"). And there already been two merges today. But if the policy should be different I can adapt.
